File: frontend/src/api/arena.ts
import { apiClient } from './client'

export interface ArenaStanding {
  team_id: number | null
  rank: number | null
  total_members: number
  battalion_points: number
  gap_to_prev: number
  arena_tier: string
  tier_label: string
  star_contributions?: { player_id: number; player_name: string; heat: number }[]
}

export interface ArenaTeamRankEntry {
  user_id: number
  nickname: string
  battalion_points: number
  arena_tier: string
  tier_label: string
}

export interface MatchArena {
  match_id: number
  team1_name: string
  team2_name: string
  status: string
  home: { team_id: number | null; name: string; power: number }
  away: { team_id: number | null; name: string; power: number }
  leader_name: string | null
  lead_points: number
  user_contributed: boolean
  frozen: boolean
}

export interface StarHeatRow {
  player_id: number
  player_name: string
  team_id?: number
  my_heat?: number
  global_heat?: number
  can_boost_today?: boolean
}

export interface MatchdayGoal {
  active: boolean
  team_id?: number
  team_name?: string | null
  progress: number
  goals: number[]
  goal_titles?: string[]
  tier_reached: number
  rewards_coins?: number[]
  my_titles?: string[]
  rally_done_today?: boolean
}

export interface TodayMatchRow {
  match_id: number
  match_date: string
  match_time: string | null
  team1_name: string
  team2_name: string
  team1_id: number | null
  team2_id: number | null
  team1_affiliation: 'primary' | 'secondary' | 'neutral'
  team2_affiliation: 'primary' | 'secondary' | 'neutral'
  team1_underdog_bonus?: number
  team2_underdog_bonus?: number
  can_cheer: boolean
  user_cheered: boolean
  user_cheer_team_id: number | null
  has_prediction?: boolean
  predict_combo_pending?: boolean
  predict_combo_after_cheer?: boolean
  cheer_block_reason: string | null
  arena: {
    home_power: number
    away_power: number
    leader_name: string | null
    lead_points: number
  }
}

export interface ArenaQuickStats {
  today_matches: number
  today_cheerable: number
  spot_remaining: number
  combo_opportunities: number
}

export interface SpotCheerStatus {
  daily_limit: number
  used_today: number
  remaining: number
  cost: number
  battalion_per_cheer: number
  slogans: string[]
  teams_today: { team_id: number; team_name: string }[]
}

export async function getArenaOverview() {
  const { data } = await apiClient.get('/api/arena/overview')
  return data as {
    standing: ArenaStanding
    next_match_arena: MatchArena | null
    my_stars: StarHeatRow[]
    matchday_goal: MatchdayGoal
    matchday_goal_secondary: MatchdayGoal
    spot_cheer: SpotCheerStatus
    card_boost_pct: number
    quick_stats: ArenaQuickStats
  }
}

export async function getTodayMatches() {
  const { data } = await apiClient.get<TodayMatchRow[]>('/api/arena/today-matches')
  return data
}

export async function getArenaTeamRank(teamId?: number, period = 'season') {
  const { data } = await apiClient.get<ArenaTeamRankEntry[]>('/api/arena/team-rank', {
    params: { team_id: teamId, period },
  })
  return data
}

export async function getMatchArena(matchId: number) {
  const { data } = await apiClient.get<MatchArena>(`/api/arena/match/${matchId}`)
  return data
}

export async function getStarHeat(scope: 'my' | 'global' = 'global', limit = 20) {
  const { data } = await apiClient.get<{ scope: string; rows: StarHeatRow[] }>('/api/arena/star-heat', {
    params: { scope, limit },
  })
  return data
}

export async function getStarAccuracy(playerId?: number) {
  const { data } = await apiClient.get('/api/arena/star-accuracy', { params: { player_id: playerId } })
  return data
}

export async function getMatchdayGoal(teamId?: number) {
  const { data } = await apiClient.get<MatchdayGoal>('/api/arena/matchday-goal', {
    params: teamId ? { team_id: teamId } : undefined,
  })
  return data
}

export async function boostStar(playerId: number) {
  const { data } = await apiClient.post('/api/arena/boost/star', { player_id: playerId })
  return data
}

export async function boostCheerExtra(matchId: number) {
  const { data } = await apiClient.post('/api/arena/boost/cheer-extra', { match_id: matchId })
  return data
}

export async function matchdayRally(teamId?: number) {
  const { data } = await apiClient.post<{
    collectible_drop?: import('./collectible').CollectibleDropResult | null
  }>('/api/arena/boost/matchday-rally', teamId != null ? { team_id: teamId } : {})
  return data
}

export async function getSpotCheerStatus() {
  const { data } = await apiClient.get<SpotCheerStatus>('/api/arena/spot-cheer')
  return data
}

export async function submitSpotCheer(teamId: number, sloganIndex: number) {
  const { data } = await apiClient.post<{
    fan_coins: number
    battalion_added: number
    slogan: string
    spot_cheer: SpotCheerStatus
  }>('/api/arena/spot-cheer', { team_id: teamId, slogan_index: sloganIndex })
  return data
}

export async function getTeamSupporters(teamId: number, period = 'season') {
  const { data } = await apiClient.get<ArenaTeamRankEntry[]>('/api/arena/team-supporters', {
    params: { team_id: teamId, period },
  })
  return data
}

File: backend/app/api/routes/arena.py
"""Arena battalion / star heat API."""

from typing import Literal

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.api.auth_deps import get_current_user, get_optional_user
from app.api.deps import get_db
from app.core.rate_limit import rate_limit_arena_spend
from app.db.models.commerce import User
from app.services.arena_service import ArenaService

router = APIRouter(prefix="/api/arena", tags=["arena"])

Period = Literal["season", "daily", "weekly"]
StarScope = Literal["global", "my"]


class BoostStarRequest(BaseModel):
    player_id: int = Field(gt=0)


class BoostCheerExtraRequest(BaseModel):
    match_id: int = Field(gt=0)


class MatchdayRallyRequest(BaseModel):
    team_id: int | None = Field(default=None, gt=0)


class SpotCheerRequest(BaseModel):
    team_id: int = Field(gt=0)
    slogan_index: int = Field(default=0, ge=0, le=20)


@router.get("/overview")
def arena_overview(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return ArenaService(db).get_overview(user)


@router.get("/team-rank")
def team_rank(
    team_id: int | None = Query(None, gt=0),
    period: Period = Query("season"),
    limit: int = Query(30, ge=1, le=100),
    db: Session = Depends(get_db),
):
    return ArenaService(db).get_team_battalion_leaderboard(team_id, period, limit)


@router.get("/team-supporters")
def team_supporters(
    team_id: int = Query(..., gt=0),
    period: Period = Query("season"),
    limit: int = Query(30, ge=1, le=100),
    db: Session = Depends(get_db),
):
    return ArenaService(db).get_team_supporter_leaderboard(team_id, period, limit)


@router.get("/spot-cheer")
def spot_cheer_status(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return ArenaService(db).get_spot_cheer_status(user)


@router.post("/spot-cheer")
def spot_cheer(body: SpotCheerRequest, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    rate_limit_arena_spend(user.id)
    return ArenaService(db).spot_cheer(user, body.team_id, body.slogan_index)


@router.get("/match/{match_id}")
def match_arena(
    match_id: int,
    user: User | None = Depends(get_optional_user),
    db: Session = Depends(get_db),
):
    return ArenaService(db).get_match_arena(match_id, user.id if user else None)


@router.get("/star-heat")
def star_heat(
    scope: StarScope = Query("global"),
    limit: int = Query(20, ge=1, le=50),
    user: User | None = Depends(get_optional_user),
    db: Session = Depends(get_db),
):
    return ArenaService(db).get_star_heat_board(user, scope=scope, limit=limit)


@router.get("/star-accuracy")
def star_accuracy(
    player_id: int | None = Query(None, gt=0),
    limit: int = Query(20, ge=1, le=50),
    db: Session = Depends(get_db),
):
    return ArenaService(db).get_star_accuracy_board(player_id, limit)


@router.get("/matchday-goal")
def matchday_goal(
    team_id: int | None = Query(None, gt=0),
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ArenaService(db).get_matchday_goal(user, team_id)


@router.get("/today-matches")
def today_matches(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return ArenaService(db).get_today_matches(user)


@router.post("/boost/star")
def boost_star(body: BoostStarRequest, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    rate_limit_arena_spend(user.id)
    return ArenaService(db).boost_star(user, body.player_id)


@router.post("/boost/cheer-extra")
def boost_cheer_extra(
    body: BoostCheerExtraRequest, user: User = Depends(get_current_user), db: Session = Depends(get_db)
):
    rate_limit_arena_spend(user.id)
    return ArenaService(db).boost_cheer_extra(user, body.match_id)


@router.post("/boost/matchday-rally")
def matchday_rally(
    body: MatchdayRallyRequest = MatchdayRallyRequest(),
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    rate_limit_arena_spend(user.id)
    return ArenaService(db).matchday_rally(user, body.team_id)
